docs/design_proposals/debug-events.md

```diff
@@ -0,0 +1,221 @@
+# Notify when debuggable containers are started and terminated 
+
+* Author(s): Brian de Alwis (@briandealwis)
+* Design Shepherd: Nick Kubala (@nkubala)
+* Date: 2019-10-24
+* Status: Reviewed
+* Related: #2211
+
+## Background
+
+Cloud Code for IntelliJ and VS Code currently launch `skaffold debug --port-forward` to start
+an application in debugging mode.  `debug` modifies the podspecs to explicitly expose the debug port used
+by the underlying language runtime via `containerPort` definitions.  Skaffold monitors the cluster
+and initiated forwards to ports on any newly-started containers, and raises events once the
+forwards are established.  The IDEs establish debug launch configurations by examining these
+port-forward events for ports with specific names to (e.g., `jdwp` for Java/JVM,
+`devtools` for NodeJS, or `dap` for connections using the _Debug Adapter Protocol_).
+When a new debuggable container is created, Skaffold will
+attempt to forward the remote debug port, which should result in a port-forward event, and thus
+spare the IDEs from having to actively watch the cluster. 
+
+But relying strictly on port-forward events carries some risk.  Although the likelihood of 
+a port-name clash is low, the IDE may require additional configuration information to establish
+a debugging connection.  For example, there may be no indication of the underlying project’s
+location in the file-system, nor the relative path in the container. 
+
+Although such information is published by `skaffold debug` via pod annotations, obtaining this
+information would require that the IDEs actively track new pods and containers.  This tracking would
+duplicate functionality already being performed by Skaffold, and also requiring this functionality to be
+re-implemented in each IDE.
+
+This design proposal recommends that Skaffold instead also watch and issue events when debuggable
+containers are started and terminated.
+
+----
+## Design
+
+This design proposes that Skaffold issue events to inform the IDEs of the following conditions:
+
+- The appearance of a container that is debuggable.  This event will provide:
+    - the namespace
+    - the pod ID
+    - the container ID
+    - the contents of the `debug.cloud.google.com/config` configuration block for the container
+      (a JSON object) including (! indicates new items)
+         - the language runtime type (`go`, `jvm`, `nodejs`, `python`)
+         - ! the image name (which should allow cross-referencing with the `skaffold.yaml`)
+         - the port(s)
```

Comment:
Can we also have the context (the DockerFile source directory) info in this event so we don't have to cross reference it in the skaffold.yaml locally?

Reply:
I don't think this event is the place for such information.  But it might make sense to have a separate query to obtain the artifact information from Skaffold (e.g., the actual artifact used, after any profile and/or kustomize/patches have been made).